Update Dockerfile and its README

Bevacizumab opened the pull request:
The original Dockerfile has issues and is not working. Ubuntu and Python versions are not specified, and the link to Psi4 is expired. This pull request fixes the Dockerfile and updates its README.md to match.

The new configuration has been tested on an Apple Silicon Mac (M4). Customization for other platforms is straightforward and can be understood easily with the updated README.md.

Changes to Dockerfile:
1. Pin versions for Ubuntu (22.04) and Python (3.12).
2. Add `WORKDIR` to mount local files.
3. Set `PATH` with `ENV`.
4. Psi4 is installed with a conda command instead of `wget`.
5. Dependencies are installed inside of a virtual environment (`fermion`).
6. Reduce the number of `RUN` commands to reduce the number of layers.

Changes to its README.md:
1. Put some details on the Dockerfile in the first paragraph of Setting up ~.
2. Add details to customization.
3. Change `docker run` command and introduce a simpler way to mount local files. So some previous explanations are deleted.

Bevacizumab committed:
Update Dockerfile and its README

mhucka commented:
> I completed the CLA last week, right after I saw the bot's comment. I've just confirmed that the CLA status is green in the "Checks" tab. It seems you might have been viewing an older notification without refreshing it.

Oh, great. Thanks for doing that.
 
> Please let me know if you have any feedback on the code changes themselves. Thank you!

Yes, I will review it.

Bevacizumab committed:
Switch to Miniforge and update REAME

Bevacizumab commented:
Hi @mhucka, I've just rebased my branch on top of the latest master. Sorry for the long wait.

mhucka commented on docker/dockerfile:
```diff
@@ -12,48 +12,46 @@
+ENV PATH="/root/conda/bin:${PATH}"
+# Set PATH for pyscf
+ENV PYTHONPATH=/root/pyscf
+
+RUN apt-get update && \
```
For DL3009, I will add a configuration setting to `.hadolint.yaml` to disable this particular check because it's just not worth the hassle.

For DL3015, go ahead and add the recommended `--no-install-recommends` flag.

For DL3008, we should probably pin the versions, although there are pros and cons to doing that. One way to find out the versions that can be used is to do roughly the following:
1. switch to the x86_64 architecture as discuss in a previous comment above
2. run the docker build as-is
3. figure out what versions of the packages were installed by the build
4. edit the file to add those vesions to the apt install command.

I don't know if it's possible with `apt-get install`, but try to see if it will let you specify just major.minor versions, not full major.minor.patch.whatever; i.e., "12.3" instead of "12.3.0.build34" etc.

Bevacizumab committed:
Version fix, architecture support, and changes to readme

Bevacizumab commented:
@mhucka Thank you for your good and detailed review comments. It helped me a lot for improving the Dockerfile. I will explain the changes in the commit.

Dockerfile
- `USER root` is deleted.
- I put `--no-install-recommends` flag for `apt-get install`.
- I thank you for adding the configuration setting.
- I fixed the package versions for `apt-get install`. Unfortunately, it requires the full strings. I've manually checked all the installed packages' versions for both arm64 and amd64 architectures. Both architectures install the same version for each package.
- I've made changes to support both linux/arm64 and linux/amd64 architectures (Psi4 only supports these two linux architectures). It is possible by putting `uname` and `uname -m` for the wget link of Miniforge.
- Additionally, I solved DL3003 warning. Each `cd` was changed to `WORKDIR`.

README
- I added information about the architecture support to help anyone who is not familiar with the concept.
- I corrected the typo `environemnt`.
- I deleted the explanation of "the default architecture".
- I reverted to CLI-based instructions. Mounting local files is a good docker practice. But some might not want to do this. So I provide instructions on how to `COPY` or not, and how to mount or not.
I split the explanation of copying local files and git repositories into a new subsection for clearance.

Bevacizumab committed:
Improved and simplified dockerfile and its readme
